kernel/syscall.c

```diff
@@ -2374,11 +2374,32 @@ int myst_syscall_bind(
 {
     long ret = 0;
     myst_fdtable_t* fdtable = myst_fdtable_current();
-    myst_sockdev_t* sd;
-    myst_sock_t* sock;
+    myst_sockdev_t *sd, *new_sd;
+    myst_sock_t *sock, *new_sock;
+    struct sockaddr* new_addr;
+    socklen_t new_addrlen;
+    bool reresolved;
 
     ECHECK(myst_fdtable_get_sock(fdtable, sockfd, &sd, &sock));
-    ret = (*sd->sd_bind)(sd, sock, addr, addrlen);
+    ECHECK(myst_sockdev_reresolve(
+        sockfd,
+        sd,
+        sock,
+        addr,
+        addrlen,
+        &reresolved,
+        &new_sd,
+        &new_sock,
+        &new_addr,
+        &new_addrlen));
+
+    if (!reresolved)
+        ret = (*sd->sd_bind)(sd, sock, addr, addrlen);
+    else
+    {
+        ret = (*new_sd->sd_bind)(new_sd, new_sock, new_addr, new_addrlen);
+        free(new_addr);
```

Comment:
When do we free the old `sock` and `addr`?

Reply:
In case of reresolution, `myst_sockdev_reresolve` closes the old `sock` object. `addr` is a buffer provided by user-space, so we don't free it. `new_addr` needs to be freed because the kernel allocated it to be able to pass the modified address to the host socket device.